ta/pkcs11/src/pkcs11_token.c

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2017-2020, Linaro Limited
+ */
+
+#include <assert.h>
+#include <pkcs11_ta.h>
+#include <string.h>
+#include <string_ext.h>
+#include <sys/queue.h>
+#include <tee_internal_api_extensions.h>
+#include <util.h>
+
+#include "pkcs11_token.h"
+#include "pkcs11_helpers.h"
+
+/* Provide 3 slots/tokens, ID is token index */
+#ifndef CFG_PKCS11_TA_TOKEN_COUNT
+#define TOKEN_COUNT		3
+#else
+#define TOKEN_COUNT		CFG_PKCS11_TA_TOKEN_COUNT
+#endif
+
+/* Static allocation of tokens runtime instances (reset to 0 at load) */
+struct ck_token ck_token[TOKEN_COUNT];
+
+/* Static allocation of tokens runtime instances */
+struct ck_token *get_token(unsigned int token_id)
+{
+	if (token_id > TOKEN_COUNT)
+		return NULL;
+
+	return &ck_token[token_id];
+}
+
+unsigned int get_token_id(struct ck_token *token)
+{
+	assert(token >= ck_token && token < &ck_token[TOKEN_COUNT]);
+
+	return token - ck_token;
+}
+
+static int pkcs11_token_init(unsigned int id)
+{
+	struct ck_token *token = init_persistent_db(id);
+
+	if (!token)
+		return 1;
+
+	if (token->state != PKCS11_TOKEN_RESET) {
+		/* Token is already in a valid state */
+		return 0;
+	}
+
+	/* Initialize the token runtime state */
+	token->state = PKCS11_TOKEN_READ_WRITE;
+	token->session_count = 0;
+	token->rw_session_count = 0;
+
+	return 0;
+}
+
+int pkcs11_init(void)
```

Comment:
I'm not that happy with these return codes. How about using `TEE_Result` instead?

Reply:
right. will fix. I'll try to remember for the next changes.